### PR TITLE
Bids cpac dev

### PR DESCRIPTION
**Pull request by joshicola**
Build/Validate/Execute model for command-line switches
Move some functions from "run.py" to "utils.py" for readability.

### CODE REVIEW

File: custom_gear_utils.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
+import os, os.path as op
+import subprocess as sp
+import sys
+from collections import OrderedDict
+import psutil
+
+import flywheel
+
+def Build_CPAC_Params(context):
```
Comment (AndysWorth) [inline custom_gear_utils.py]: Make name generic
```suggestion
def Build_Params(context):
```

File: custom_gear_utils.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    # use Ordered Dictionary to keep the order created.
+    # Default in Python 3.6 onward
+    Params=OrderedDict()
+
+    if context.get_input_path('pipeline_file'):
+        Params['pipeline_file'] = context.get_input_path('pipeline_file')
+
+    # This "black_list" list will skip keys if present in context.Custom_Dict
```
Comment (AndysWorth) [inline custom_gear_utils.py:17]: 👍

File: custom_gear_utils.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                if len(key) == 1:
+                    Params[key] = config[key]
+                else:
+                    # if the key-value is zero, we skip and use the defaults
+                    if config[key] > 0:
+                        Params[key] = config[key]
+
+    context.Custom_Dict['cpac_params'] = Params
+
+def Validate_CPAC_Params(context):
```
Comment (AndysWorth) [inline custom_gear_utils.py]: Make generic
```suggestion
def Validate_Params(context):
```

File: custom_gear_utils.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            else:
+                # If Param not boolean, without value include without value
+                # (e.g. '--key'), else include value (e.g. '--key=value')
+                if len(str(ParamList[key])) == 0:
+                    command.append('--' + key)
+                else:
+                    command.append('--' + key + '=' + str(ParamList[key]))
+    return command
+
+def Execute_CPAC_Params(context, dry_run=True):
```
Comment (AndysWorth) [inline custom_gear_utils.py]: Make generic
```suggestion
def Execute_Params(context, dry_run=True):
```

**Review by AndysWorth — APPROVED** [review]
At first I thought that all of the "util" type thing should be generic but now I see why you separated custom_gear_utils.py from utils.py.
Looks good!

File: run.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import os, os.path as op
+import subprocess as sp
+import sys
+import zipfile
+from collections import OrderedDict
+import psutil
+
+import flywheel
+from utils import *
+from custom_gear_utils import *
```
Comment (kalebdfischer) [inline run.py]: Best practice to avoid `from package import *`. `import package` is more readable and avoids import conflicts.

File: utils.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+"""
+This file contains various common utilities that facilitate gear excecution
+"""
+import logging
+import sys
+
+def escape_shell_chars(path):
```
Comment (kalebdfischer) [inline utils.py:8]: Could replace with:
```
import re
path = re.sub('[^0-9a-zA-Z]+', '_', path)
```
for quicker execution and avoiding long repeats of underscores.

File: custom_gear_utils.py (hunk 1 above)
Comment (joshicola) [reply custom_gear_utils.py]: RE: Remove _CPAC:
Generally, a good idea here.

In another gear, hcp-struct, we have three Build/Validate/Execute blocks that need to be named separately.

With only one block here, I can see better readability.

File: utils.py (hunk 6 above)
Comment (joshicola) [reply utils.py:8]: Thanks, Kaleb,
With some modifications, that will be much faster and take into consideration a full path (e.g. /path/to/my/file.nii.gz) for shell-safe characters that excluding all non-alphanumeric would miss.

**Pull request merged**